### PR TITLE
Ms.fix coin selection

### LINKED ISSUE #12226: [Bug] Repeated timeout when trying to send XCH
### What happened?

Attempting to send ~2 XCH from one wallet to another on Win10 Chia 1.4 and getting repeated timeouts on the Tx. 
EDIT: Amount accuracy 

Never had trouble with this on past versions. Any suggestions? Thanks.

### Version

1.4

### What platform are you using?

Windows

### What ui mode are you using?

GUI

### Relevant log output or stacktrace

```shell
2022-07-02T16:xxxxxx wallet chia.rpc.rpc_server        : WARNING  Error while handling message: Traceback (most recent call last):
  File "chia\rpc\rpc_server.py", line 229, in safe_handle
  File "chia\rpc\rpc_server.py", line 220, in ws_api
  File "chia\rpc\wallet_rpc_api.py", line 838, in send_transaction
  File "chia\wallet\wallet.py", line 432, in generate_signed_transaction
  File "chia\wallet\wallet.py", line 315, in _generate_unsigned_transaction
  File "chia\wallet\wallet.py", line 267, in select_coins
  File "chia\wallet\coin_selection.py", line 90, in select_coins
  File "chia\wallet\coin_selection.py", line 119, in select_smallest_coin_over_target
ValueError: Unable to select coins for this transaction. Try sending a smaller amount
```

### PR DESCRIPTION
**Pull request by mariano54**
Two bugs :
1. `test_smallest_coin_over_amount` did not work properly when all coins where smaller than the amount.
2. In the cases where knapsack found a solution but it was too large, we fell back to selecting the "largest coin over the amount" as in issue 1, but instead we should be adding up the largest coins that are < the target amount first, because we might not have any large coins.

It also fixes a performance issue with knapsack, where it keeps searching even if we are selecting more coins than we can actually select.
Old performance with 200k coins:  60 seconds. New: 0.78 seconds.

A bunch of tests are added to test all the functionality.

### CODE REVIEW

**Comment by emlowe** [comment]
`valid_spendable_coins` is already sorted in descending order when passed into `select_smallest_coin_over_target` so I don't see why you are sorting it again. Same with `sum_largest_coins` - the `smaller_coin` list is already sorted in the order you need. The prior implementation logic was actually correct but with confusing parameter names.

**Comment by jack60612** [comment]
> `valid_spendable_coins` is already sorted in descending order when passed into `select_smallest_coin_over_target` so I don't see why you are sorting it again. Same with `sum_largest_coins` - the `smaller_coin` list is already sorted in the order you need. The prior implementation logic was actually correct but with confusing parameter names.

There is a bug, but it is really hard to see, so mariano simplified the logic and fixed the bug.
The increase in time is under a couple of milliseconds and it is worth the really easy to read logic in my opinion

**Review by jack60612 — CHANGES_REQUESTED** [review]
Just some minor test changes

**Comment by jack60612** [comment]
re running ci

**Comment by mariano54** [comment]
> `valid_spendable_coins` is already sorted in descending order when passed into `select_smallest_coin_over_target` so I don't see why you are sorting it again. Same with `sum_largest_coins` - the `smaller_coin` list is already sorted in the order you need. The prior implementation logic was actually correct but with confusing parameter names.

There was a bug when `smaller_coin_amount == len(valid_spendable_coin_list)`. I refactored that method so it's clear what it does now.

The reason I sorted within the method is that it makes the API simpler: it now no longer depends on a specific sorting. I thought that this API is harder to mess up for callers. For reference, sorting 200,000 coins in a worst case ordering takes about 0.4 seconds. This is probably OK when making a transaction, but I can see an argument either way.

File: tests/wallet/test_coin_selection.py
Comment (mariano54) [reply tests/wallet/test_coin_selection.py:174]: Are you sure that it's exactly the same? I'm not sure about that, I rather keep both for safety

**Comment by mariano54** [comment]
@jack60612 @emlowe , take a look at the new changes.
Knapsack perf is fixed, and I reverted the sorting for more performance

File: chia/wallet/coin_selection.py
Comment (jack60612) [inline chia/wallet/coin_selection.py:88]: `len(coin_set) > max_num_coins` cant happen

File: chia/wallet/coin_selection.py
Comment (jack60612) [inline chia/wallet/coin_selection.py:122]: remove comment

**Review by jack60612 — DISMISSED** [review]
lgtm

**Review by jack60612 — DISMISSED** [review]
lgtm

**Comment by jack60612** [comment]
once he fixes the lint issues i still approve it, just merge it even if i dont have the chance to re approve

**Review by jack60612 — APPROVED** [review]
lgtm

File: chia/wallet/coin_selection.py
Comment (emlowe) [inline chia/wallet/coin_selection.py:103]: In the interest of consistency - should this be the same error text as above

` f"Transaction of {amount} mojo would use more than {max_num_coins} coins. Try sending a smaller amount"`

**Review by emlowe — APPROVED** [review]
Approved - minor comment on error string, but not gating

**Pull request merged**
